keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1328 @@
+
+Scalability: In the limit, ClusterTrustBundle objects will be used by
+every pod in the cluster, which will require watches from all Kubelets.  When
+they are updated, workloads will need to receive the updates fairly quickly
+(within 5 minutes across the whole cluster), to accommodate emergency rotation
+of trust anchors for a private CA.
+
+Security: Should individual trust anchor set entries designate an OSCP endpoint
+to check for certificate revociation?  Or should we require the URL to be
+embedded in the issued certificates?
+
+## Design Details
+
+### ClusterTrustBundle Object
+
+The ClusterTrustBundle object allows signers to publish their trust anchors for
+consumption by workloads, whether directly, via Kubelet, or otherwise.
+
+ClusterTrustBundle objects may optionally be associated with a signer using the
+`.spec.signerName` field.  Setting the `.spec.signerName` field enables
+additional admission control and access control for the ClusterTrustBundle
+object, as detailed in the [Access Control](#access-control) section.
+
+While multiple ClusterTrustBundle objects may be associated with a signer, each
+is a completely independent expression of the roots of trust for the signer.  No
+effort is made to unify trust anchors from different ClusterTrustBundle objects.
+The primary use case for multiple ClusterTrustBundle objects for a single signer
+is [canarying](#canarying-changes-to-a-clustertrustbundle).
+
+All of the new objects and fields described in this section are gated by the
+ClusterTrustBundle kube-apiserver feature gate.  Unless the feature gate is
+enabled, usage of these alpha objects and fields will be rejected by
+kube-apiserver.
+
+#### API Object Definition
+
+```go
+// ClusterTrustBundle packages up trust anchors.
+//
+// Each ClusterTrustBundle object is optionally owned by a particular signer, which
+// updates the trust anchors as the roots of trust for that signer change.
+//
+// A ClusterTrustBundle object wraps up multiple trust anchor sets, each of
+// which contains one or more trust anchors (individual
+// certificates).  Most use cases will only require a single trust
+// anchor set, conventionally named `default`.  Multiple uniquely-named trust
+// anchor sets will likely only be required in mutual TLS use cases with
+// federation between multiple meshes.
+type ClusterTrustBundle struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // The spec of the object.
+    Spec ClusterTrustBundleSpec `json:"spec"`
+
+    // The status of the object.
+    Status ClusterTrustBundleStatus `json:"status"`
+}
+
+// ClusterTrustBundleSpec is the spec subfield of a ClusterTrustBundle object.
+type ClusterTrustBundleSpec struct {
+  // The name of the associated signer.
+  //
+  // +optional
+  SignerName string `json:"signerName,omitempty"`
+}
+
+// ClusterTrustBundleStatus is the status subfield of a ClusterTrustBundle object.
+type ClusterTrustBundleStatus struct {
```

Comment:
I'm not convinced that Status makes sense for this object. What is kubelet supposed to do if a ClusterTrustBundle is not Ready? If a ClusterTrustBundle isn't ready, why doesn't the signer just update it?

Reply:
Kubelet is not supposed to project certificates from a cluster trust bundle that is not Ready.  This is primarily meant to cover integrations with cloud-backed CAs.  In that case, the signer implementation has to retrieve the PEM data from the cloud CA on startup, and needs a way to signal to consumers that the ClusterTrustBundle is not ready to use.

Reply:
I decided to just put a single pem set in the spec field, and remove the status subobject.  The safety check I was desiring can be handled just by making sure a signer implementation never creates an empty ClusterTrustBundle object.